[Vulkan] Add reflection_pad2d to Vulkan

SS-JIA opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53604 [Vulkan] Add reflection_pad2d to Vulkan**

Differential Revision: [D27098310](https://our.internmc.facebook.com/intern/diff/D27098310)

Commit:
[Vulkan] Add reflection_pad2d to Vulkan

[ghstack-poisoned]

Commit:
Update on "[Vulkan] Add reflection_pad2d to Vulkan"




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/reflection_pad2d.glsl:
```diff
@@ -0,0 +1,31 @@
+    const int xoff_pre  = 2*max(uBlock.pad.x - pos.x, 0);
+    const int xoff_post = 2*max(pos.x - (uBlock.size.x - 1 - uBlock.pad.y), 0);
+    const int yoff_pre  = 2*max(uBlock.pad.z - pos.y, 0);
+    const int yoff_post = 2*max(pos.y - (uBlock.size.y - 1 - uBlock.pad.w), 0);
```
I think these statements can be packed in terms of vectors as opposed to scalars.  For instance

```
const int2 xoff = 2 * max(uBlock.pad.xz - pos.xy, 0);
const int2 yoff = 2 * max(pos.xy - (uBlock.size.xy - 1 - uBlock.pad.yw), 0);
```

and even then you might be able to refactor it (with some thought - not obvious to me now) to

```
const int4 offset = 2 * max( ... );
```

SS-JIA replied:
you're absolutely right, can definitely make this part more compact.

AshkanAliabadi commented on aten/src/ATen/native/vulkan/glsl/reflection_pad2d.glsl:
```diff
@@ -0,0 +1,31 @@
+    ivec3 inpos = ivec3(
+        pos.x - uBlock.pad.x + xoff_pre - xoff_post,
+        pos.y - uBlock.pad.z + yoff_pre - yoff_post,
+        pos.z);
```
ivec3 has a constructor for (ivec2, int) in case you can manage to combine the first two parameters into a vector.

```
const ivec3 inpos = ivec3(pos.xz - uBlock.pad.xz + ?, pos.z)
```

AshkanAliabadi commented on aten/src/ATen/native/vulkan/ops/Padding.cpp:
```diff
@@ -0,0 +1,109 @@
+  const Tensor self = self_arg.is_vulkan() ? self_arg : self_arg.vulkan();
+  const vTensor& v_self = convert(self);
+
+  std::vector<int64_t> output_size(input_dim);
```
Please use c10::SmallVector<int64_t, 4> (or whatever default size you think is best).  That will save a heap allocation since the first 4 (or the number of your choosing) elements are on the stack.

On the other hand if you want this function to work (or know that this is the only combination that makes sense) only for 4d tensors and below you can use std::array.

Commit:
Update on "[Vulkan] Add reflection_pad2d to Vulkan"


Differential Revision: [D27098310](https://our.internmc.facebook.com/intern/diff/D27098310)

[ghstack-poisoned]

Commit:
Update on "[Vulkan] Add reflection_pad2d to Vulkan"


Differential Revision: [D27098310](https://our.internmc.facebook.com/intern/diff/D27098310)

[ghstack-poisoned]
